File: dt_brightlifecare_customization/public/py/stock_entry.py
import frappe
import time




def on_submit(doc, method):
    make_se_for_code_to_code_transfer(doc, method)
    make_serial_and_barcode_for_fg_item(doc, method)
    
    submit_stock_entry_with_qi(doc, method=None)
    issued_qty_calculation_in_wo(doc, method)
    returned_qty_calculation_in_wo(doc, method)
    consumed_qty_calculation_in_wo(doc, method)





def submit_stock_entry_with_qi(doc, method=None):
    """
    After Stock Entry submission, wait for Serial & Batch Bundle creation
    then create QI for finished items if BOM requires quality inspection.
    """
    # Only for Manufacture entries
    if doc.stock_entry_type != "Manufacture":
        return

    # Ensure BOM is linked
    if not doc.bom_no:
        return

    # Check custom checkbox on BOM
    bom_flag = frappe.db.get_value("BOM", doc.bom_no, "custom_quality_inspection")
    if not bom_flag:
        return

    # Wait 10 seconds (to ensure bundles created)
    time.sleep(10)

    se = frappe.get_doc("Stock Entry", doc.name)
    inspected_by = frappe.session.user

    for item in se.items:
        # Only finished items
        if not item.is_finished_item:
            continue

        frappe.log_error(f"Processing finished item {item.item_code}, bundle: {item.serial_and_batch_bundle}", "QI Debug")

        # Avoid duplicate QI
        if not frappe.db.exists({
            "doctype": "Quality Inspection",
            "reference_type": "Stock Entry",
            "reference_name": se.name,
            "item_code": item.item_code
        }):
            sample_size = frappe.db.get_value("Item", item.item_code, "sample_quantity") or 0

            batch_no = None
            if item.serial_and_batch_bundle:
                try:
                    bundle = frappe.get_doc("Serial and Batch Bundle", item.serial_and_batch_bundle)
                    if bundle.entries:
                        batch_no = bundle.entries[0].batch_no
                except Exception as e:
                    frappe.log_error(f"Error fetching batch for item {item.item_code}: {e}", "QI Debug")

            # Create QI in Draft
            qi = frappe.get_doc({
                "doctype": "Quality Inspection",
                "inspection_type": "Incoming",
                "reference_type": "Stock Entry",
                "reference_name": se.name,
                "item_code": item.item_code,
                "company": se.company,
                "status": "Accepted",   # keep in draft state
                "sample_size": sample_size,
                "inspected_by": inspected_by,
                "inspection_date": se.posting_date,
                "batch_no": batch_no,
                "serial_and_batch_bundle": item.serial_and_batch_bundle
            })
            qi.insert()
            frappe.log_error(f"QI created for item {item.item_code} with batch {batch_no}", "QI Debug")







def make_serial_and_barcode_for_fg_item(doc, method):
    """
    When a Manufacture Stock Entry is submitted:
    - Check if linked to Work Order
    - Identify finished good item (t_warehouse present)
    - If item requires serial numbers, create Serial Number and Barcode doc
    - Populate child table with all serial numbers generated
    """
    if doc.stock_entry_type != "Manufacture" or not doc.work_order:
        return

    for item in doc.items:
        if not item.t_warehouse:
            continue

        has_serial_no = frappe.get_value("Item", item.item_code, "has_serial_no")
        if not has_serial_no:
            continue

        # Fetch all serial numbers linked to this Stock Entry and item
        serial_nos = frappe.get_all(
            "Serial No",
            filters={"item_code": item.item_code, "work_order": doc.work_order},
            pluck="name"
        )

        if not serial_nos:
            continue

        # Create parent document
        snb = frappe.get_doc({
            "doctype": "Serial Number and Barcode",
            "item_code": item.item_code,
            "work_order": doc.work_order,
            "stock_entry": doc.name
        })

        # Add child rows for each serial number
        for sn in serial_nos:
            snb.append("serial_no_and_barcode_detail", {
                "serial_number": sn
            })

        snb.insert()
        # frappe.db.commit()







# import frappe

# def on_cancel(doc, method):
#     """
#     When a Stock Entry is cancelled, remove Serial Number and Barcode docs
#     linked to its Work Order and finished good items.
#     """
#     if doc.stock_entry_type != "Manufacture" or not doc.work_order:
#         return

#     for item in doc.items:
#         if not item.t_warehouse:
#             continue

#         has_serial_no = frappe.get_value("Item", item.item_code, "has_serial_no")
#         if not has_serial_no:
#             continue

#         # Find all Serial Number and Barcode docs created for this WO + Item
#         snb_list = frappe.get_all(
#             "Serial Number and Barcode",
#             filters={"item_code": item.item_code, "work_order": doc.work_order, "stock_entry": doc.name},
#             pluck="name"
#         )

#         for snb in snb_list:
#             frappe.delete_doc("Serial Number and Barcode", snb)

    # frappe.db.commit()














def make_se_for_code_to_code_transfer(doc, method):
    if doc.stock_entry_type != "Manufacture":
        return

    for item in doc.items:
        # Check if code-to-code transfer is enabled for this item
        ctc = frappe.db.get_value("Item", item.item_code, "custom_code_to_code_transfer")
        ctc_item = frappe.db.get_value("Item", item.item_code, "custom_code_to_code_transfer_item")

        if ctc and ctc_item:
            try:
                CtC_SE = frappe.get_doc({
                    "doctype": "Stock Entry",
                    "stock_entry_type": "Code To Code Transfer",
                    "company": doc.company,
                    "posting_date": doc.posting_date,
                    "posting_time": doc.posting_time,
                    "set_posting_time": 1
                })

                # Outgoing (from t_warehouse of manufacture entry)
                CtC_SE.append("items", {
                    "s_warehouse": item.t_warehouse,
                    "item_code": item.item_code,
                    "qty": item.qty
                })

                # Incoming (into t_warehouse of manufacture entry)
                CtC_SE.append("items", {
                    "t_warehouse": item.t_warehouse,
                    "item_code": ctc_item,
                    "qty": item.qty
                })

                # Save and submit
                CtC_SE.insert()
                CtC_SE.submit()

                frappe.msgprint(f"Code to Code Transfer created: {CtC_SE.name}")

            except Exception as e:
                frappe.log_error(frappe.get_traceback(), "Error in make_se_for_code_to_code_transfer")
                frappe.msgprint(f"Error while creating Code to Code Transfer: {str(e)}")


import frappe

@frappe.whitelist()
def print_dispensing_slips_preview(docname):
   doc = frappe.get_doc("Stock Entry", docname)

   # Just get HTML of your custom print format
   html = frappe.get_print(
       doctype="Stock Entry",
       name=doc.name,
       print_format="DT Dispensing Slips for RM",
       doc=doc,
       as_pdf=False
   )

   return html








def before_save(doc, method):
    for item in doc.items:
        if item.ste_detail:
            se_doc = frappe.get_doc("Stock Entry Detail", item.ste_detail)

            if se_doc.qty and item.qty > se_doc.qty:
                frappe.throw("Quanity must be less than or equals to linked Stock Entry quanity")





# @frappe.whitelist()
# def remove_unwanted_rows(row_name):
#     source = frappe.get_doc("Stock Entry Detail", row_name)
#     source.delete()







import frappe

def issued_qty_calculation_in_wo(doc, method):
    if doc.stock_entry_type != "Transfer to Manufacturing Machine Setup":
        return

    if not doc.work_order:
        return

    # Fetch submitted Work Order
    wo = frappe.get_doc("Work Order", doc.work_order)

    if wo.docstatus != 1:
        return  # Safety check

    for se_item in doc.items:
        if not se_item.item_code or not se_item.qty:
            continue

        wo.append("custom_machine_setup_inventory_detail", {
            "item_code": se_item.item_code,
            "issued_qty": se_item.qty
        })

    # Save submitted document safely
    wo.save()







def returned_qty_calculation_in_wo(doc, method):
    if doc.stock_entry_type != "Machine Setup Return":
        return

    if not doc.work_order:
        return

    # Fetch submitted Work Order
    wo = frappe.get_doc("Work Order", doc.work_order)

    if wo.docstatus != 1:
        return

    # Build item_code → row map for quick lookup
    wo_items_map = {
        d.item_code: d
        for d in wo.custom_machine_setup_inventory_detail
        if d.item_code
    }

    for se_item in doc.items:
        if not se_item.item_code or not se_item.qty:
            continue

        if se_item.item_code in wo_items_map:
            # 🔁 Item exists → add qty
            row = wo_items_map[se_item.item_code]
            row.returned_qty = (row.returned_qty or 0) + se_item.qty
        else:
            # ➕ Item not present → add new row
            wo.append("custom_machine_setup_inventory_detail", {
                "item_code": se_item.item_code,
                "returned_qty": se_item.qty
            })

    # Save submitted document safely
    wo.save()








def consumed_qty_calculation_in_wo(doc, method):
    if doc.stock_entry_type != "Machine Setup Consumption Entry":
        return

    if not doc.work_order:
        return

    # Fetch submitted Work Order
    wo = frappe.get_doc("Work Order", doc.work_order)

    if wo.docstatus != 1:
        return

    # Build item_code → row map for quick lookup
    wo_items_map = {
        d.item_code: d
        for d in wo.custom_machine_setup_inventory_detail
        if d.item_code
    }

    print("\n\n\n",wo_items_map,"\n\n\n")

    for se_item in doc.items:
        if not se_item.item_code or not se_item.qty:
            continue

        if se_item.item_code in wo_items_map:
            # 🔁 Item exists → add qty
            row = wo_items_map[se_item.item_code]
            row.consumed_qty = (row.consumed_qty or 0) + se_item.qty
            print("\n\n\n",se_item.item_code, row.consumed_qty,"\n\n\n")
        else:
            # ➕ Item not present → add new row
            wo.append("custom_machine_setup_inventory_detail", {
                "item_code": se_item.item_code,
                "consumed_qty": se_item.qty
            })

    # Save submitted document safely
    wo.save()

File: dt_brightlifecare_customization/public/js/stock_entry.js
frappe.ui.form.on('Stock Entry', {
    refresh: function(frm) {
        frm.fields_dict.items.grid.get_field('uom').get_query = function(doc, cdt, cdn) {
            const item_code = frappe.get_doc(cdt, cdn).item_code;
            return {
                query: 'dt_brightlifecare_customization.public.py.uom_filter.uom_filter_condition',
                filters: {
                    item_code: item_code
                }
            };
        };


        if (frm.doc.docstatus === 1 && frm.doc.stock_entry_type == "Transfer to Manufacturing Machine Setup") {

            frm.add_custom_button('Create Reversal SE', function () {
                let new_se = frappe.model.get_new_doc('Stock Entry');

                new_se.stock_entry_type = "Machine Setup Return";
                new_se.work_order = frm.doc.work_order;
                new_se.from_bom = 0;
                new_se.against_stock_entry = frm.doc.name;

                // Map items from current SE
                let reversed_items = frm.doc.items.map(item => {
                    return {
                        item_code: item.item_code,
                        item_name: item.item_name,
                        qty: 0,
                        uom: item.uom,
                        conversion_factor: item.conversion_factor,
                        basic_rate: item.basic_rate,
                        s_warehouse: item.t_warehouse,
                        t_warehouse: item.s_warehouse,
                        against_stock_entry: frm.doc.name,
                        ste_detail: item.name,
                        batch_no: item.batch_no,
                    };
                });

                frappe.set_route('Form', 'Stock Entry', new_se.name).then(() => {
                    let target_frm = cur_frm;

                    if (target_frm && target_frm.doctype === 'Stock Entry') {

                        // Clear items table first
                        target_frm.clear_table('items');

                        reversed_items.forEach(d => {
                            let row = target_frm.add_child('items');

                            // Suppress batch/serial popup by setting __run_batch_popup = false
                            row.__run_batch_popup = false;

                            // Set values programmatically
                            Object.keys(d).forEach(field => {
                                row[field] = d[field];
                            });
                        });

                        target_frm.refresh_field('items');
                        target_frm.set_value('from_bom', 0);
                    }
                });
            });

        }

    }
});


frappe.ui.form.on('Stock Entry', {
    custom_validate_batch: function (frm) {
        // This runs when you click your existing button
        open_camera_dialog(frm);
    },

    refresh: function(frm) {

        frm.fields_dict.custom_parameters.grid.wrapper
            .find('.grid-add-row, .grid-remove-rows').hide();

        // Also disable the ability to add/delete via keyboard shortcuts or API
        frm.fields_dict.custom_parameters.grid.cannot_add_rows = true;
        frm.fields_dict.custom_parameters.grid.cannot_delete_rows = true;


        if(frm.doc.stock_entry_type === "Material Transfer for Manufacture") {
            frm.add_custom_button(__('Print Dispensing Tags'), function() {
                frappe.call({
                    method: "dt_brightlifecare_customization.public.py.stock_entry.print_dispensing_slips_preview",
                    args: { docname: frm.doc.name },
                    callback: function(r) {
                        if(r.message){
                            let w = window.open('', '_blank');
                            w.document.write(`
                                <html>
                                    <head>
                                        <title>Dispensing Slips - ${frm.doc.name}</title>
                                        <style>
                                            @media print {
                                                .page-break { page-break-after: always; }
                                            }
                                        </style>
                                    </head>
                                    <body>
                                        ${r.message}
                                    </body>
                                </html>
                            `);
                            w.document.close();
                        }
                    }
                });
            });
        };
   }
});



// function open_camera_dialog(frm) {
//     // Load jsQR (lightweight QR decoder)
//     frappe.require("https://cdn.jsdelivr.net/npm/jsqr@1.4.0/dist/jsQR.min.js", function () {

//         let d = new frappe.ui.Dialog({
//             title: "Scan QR Code",
//             fields: [
//                 { fieldtype: "HTML", fieldname: "camera_area" }
//             ],
//             primary_action_label: "Close",
//             primary_action: function () {
//                 stopCamera();
//                 d.hide();
//             }
//         });

//         d.show();

//         const wrapper = d.get_field("camera_area").$wrapper;
//         const vidId = "se-camera-" + frappe.utils.get_random(6);
//         const canvasId = "se-canvas-" + frappe.utils.get_random(6);

//         wrapper.html(`
//             <div style="text-align:center;">
//                 <video id="${vidId}" autoplay playsinline style="max-width:100%; height:auto; border:1px solid #ccc; border-radius:6px;"></video>
//                 <canvas id="${canvasId}" style="display:none;"></canvas>
//             </div>
//             <div style="text-align:center; margin-top:8px;">
//                 <small>Show a QR code in front of the camera to check.</small>
//             </div>
//         `);

//         const video = wrapper.find(`#${vidId}`)[0];
//         const canvas = wrapper.find(`#${canvasId}`)[0];
//         const ctx = canvas.getContext("2d");

//         let currentStream = null;
//         let scanning = true;

//         function stopCamera() {
//             scanning = false;
//             if (currentStream) {
//                 currentStream.getTracks().forEach(t => t.stop());
//                 currentStream = null;
//             }
//         }

//         const modal = d.$wrapper.find('.modal');
//         modal.on('hidden.bs.modal', () => {
//             stopCamera();
//         });

//         if (!navigator.mediaDevices || !navigator.mediaDevices.getUserMedia) {
//             frappe.msgprint("Camera API not supported in this browser.");
//             return;
//         }

//         navigator.mediaDevices.getUserMedia({ video: { facingMode: "environment" } })
//             .then(stream => {
//                 currentStream = stream;
//                 video.srcObject = stream;
//                 video.play();

//                 const tick = () => {
//                     if (!scanning) return;
//                     if (video.readyState === video.HAVE_ENOUGH_DATA) {
//                         canvas.height = video.videoHeight;
//                         canvas.width = video.videoWidth;
//                         ctx.drawImage(video, 0, 0, canvas.width, canvas.height);
//                         const imageData = ctx.getImageData(0, 0, canvas.width, canvas.height);
//                         const code = jsQR(imageData.data, imageData.width, imageData.height);
//                         if (code) {
//                             stopCamera();
//                             d.hide();

//                             // Extract batch number (handles both plain and full URLs)
//                             let scanned_value = code.data.trim();
//                             let batch_no = scanned_value.split("/").pop();

//                             let matched = false;
//                             let changed = false;

//                             // Iterate through Stock Entry items table
//                             (frm.doc.items || []).forEach(row => {
//                                 if (row.batch_no && row.batch_no === batch_no) {
//                                     if (row.custom_batch_validation !== "Matched") {
//                                         frappe.model.set_value(row.doctype, row.name, "custom_batch_validation", "Matched");
//                                         changed = true;
//                                     }
//                                     matched = true;
//                                 }
//                             });

//                             frm.refresh_field("items");

//                             if (!matched) {
//                                 frappe.msgprint(`❌ No matching batch found for: ${batch_no}`);
//                             } else {
//                                 frappe.show_alert({ message: `✅ Batch ${batch_no} matched and updated!`, indicator: "green" });

//                                 // Auto-save only if something changed
//                                 if (changed) {
//                                     frm.save()
//                                         .then(() => {
//                                             frappe.show_alert({ message: "💾 Stock Entry saved successfully!", indicator: "blue" });
//                                         })
//                                         .catch(() => {
//                                             frappe.msgprint("⚠️ Failed to auto-save the Stock Entry.");
//                                         });
//                                 }
//                             }
//                             return;
//                         }
//                     }
//                     requestAnimationFrame(tick);
//                 };
//                 requestAnimationFrame(tick);
//             })
//             .catch(err => {
//                 frappe.msgprint("Unable to access camera: " + err.message);
//             });
//     });
// }





function open_camera_dialog(frm) {
    // Load jsQR (lightweight QR decoder)
    frappe.require("https://cdn.jsdelivr.net/npm/jsqr@1.4.0/dist/jsQR.min.js", function () {

        let d = new frappe.ui.Dialog({
            title: "Scan QR Code",
            fields: [
                { fieldtype: "HTML", fieldname: "camera_area" }
            ],
            primary_action_label: "Close",
            primary_action: function () {
                stopCamera();
                d.hide();
            }
        });

        d.show();

        const wrapper = d.get_field("camera_area").$wrapper;
        const vidId = "se-camera-" + frappe.utils.get_random(6);
        const canvasId = "se-canvas-" + frappe.utils.get_random(6);

        wrapper.html(`
            <div style="text-align:center;">
                <video id="${vidId}" autoplay playsinline
                    style="max-width:100%; height:auto; border:1px solid #ccc; border-radius:6px;">
                </video>
                <canvas id="${canvasId}" style="display:none;"></canvas>
            </div>
            <div style="text-align:center; margin-top:8px;">
                <small>Show a QR code in front of the camera to check.</small>
            </div>
        `);

        const video = wrapper.find(`#${vidId}`)[0];
        const canvas = wrapper.find(`#${canvasId}`)[0];
        const ctx = canvas.getContext("2d");

        let currentStream = null;
        let scanning = true;

        function stopCamera() {
            scanning = false;
            if (currentStream) {
                currentStream.getTracks().forEach(t => t.stop());
                currentStream = null;
            }
        }

        const modal = d.$wrapper.find('.modal');
        modal.on('hidden.bs.modal', () => {
            stopCamera();
        });

        if (!navigator.mediaDevices || !navigator.mediaDevices.getUserMedia) {
            frappe.msgprint("Camera API not supported in this browser.");
            return;
        }

        navigator.mediaDevices.getUserMedia({ video: { facingMode: "environment" } })
            .then(stream => {
                currentStream = stream;
                video.srcObject = stream;
                video.play();

                const tick = () => {
                    if (!scanning) return;

                    if (video.readyState === video.HAVE_ENOUGH_DATA) {
                        canvas.height = video.videoHeight;
                        canvas.width = video.videoWidth;
                        ctx.drawImage(video, 0, 0, canvas.width, canvas.height);

                        const imageData = ctx.getImageData(0, 0, canvas.width, canvas.height);
                        const code = jsQR(
                            imageData.data,
                            imageData.width,
                            imageData.height
                        );

                        if (code) {
                            stopCamera();
                            d.hide();

                            // ✅ Extract batch number from QR
                            let scanned_value = code.data.trim();
                            let batch_no = scanned_value.split("/").pop();

                            let matched = false;
                            let changed = false;
                            let pending = 0;

                            (frm.doc.items || []).forEach(row => {
                                if (!row.serial_and_batch_bundle) return;

                                pending++;

                                frappe.model.with_doc(
                                    "Serial and Batch Bundle",
                                    row.serial_and_batch_bundle,
                                    function () {
                                        let bundle = frappe.model.get_doc(
                                            "Serial and Batch Bundle",
                                            row.serial_and_batch_bundle
                                        );

                                        if (bundle && bundle.entries) {
                                            bundle.entries.forEach(entry => {
                                                if (entry.batch_no === batch_no) {
                                                    matched = true;

                                                    if (row.custom_batch_validation !== "Matched") {
                                                        frappe.model.set_value(
                                                            row.doctype,
                                                            row.name,
                                                            "custom_batch_validation",
                                                            "Matched"
                                                        );
                                                        changed = true;
                                                    }
                                                }
                                            });
                                        }

                                        pending--;

                                        // ✅ Decide result ONLY after all bundles checked
                                        if (pending === 0) {
                                            frm.refresh_field("items");

                                            if (!matched) {
                                                frappe.msgprint(
                                                    `❌ No matching batch found for: ${batch_no}`
                                                );
                                            } else {
                                                frappe.show_alert({
                                                    message: `✅ Batch ${batch_no} matched and updated!`,
                                                    indicator: "green"
                                                });

                                                if (changed) {
                                                    if (frm.doc.docstatus === 0) {
                                                        // Draft → normal save
                                                        frm.save();
                                                    } else if (frm.doc.docstatus === 1) {
                                                        // Submitted → Update After Submit
                                                        frm.save('Update');
                                                    }
                                                }


                                            }
                                        }
                                    }
                                );
                            });

                            return;
                        }
                    }

                    requestAnimationFrame(tick);
                };

                requestAnimationFrame(tick);
            })
            .catch(err => {
                frappe.msgprint("Unable to access camera: " + err.message);
            });
    });
}




// opens camera and diaply msg when QR code shown 
// function open_camera_dialog(frm) {
//     // Load jsQR (lightweight QR decoder)
//     frappe.require("https://cdn.jsdelivr.net/npm/jsqr@1.4.0/dist/jsQR.min.js", function () {

//         let d = new frappe.ui.Dialog({
//             title: "Scan QR Code",
//             fields: [
//                 { fieldtype: "HTML", fieldname: "camera_area" }
//             ],
//             primary_action_label: "Close",
//             primary_action: function () {
//                 stopCamera();
//                 d.hide();
//             }
//         });

//         d.show();

//         const wrapper = d.get_field("camera_area").$wrapper;
//         const vidId = "se-camera-" + frappe.utils.get_random(6);
//         const canvasId = "se-canvas-" + frappe.utils.get_random(6);

//         wrapper.html(`
//             <div style="text-align:center;">
//                 <video id="${vidId}" autoplay playsinline style="max-width:100%; height:auto; border:1px solid #ccc; border-radius:6px;"></video>
//                 <canvas id="${canvasId}" style="display:none;"></canvas>
//             </div>
//             <div style="text-align:center; margin-top:8px;">
//                 <small>Show a QR code in front of the camera to test.</small>
//             </div>
//         `);

//         const video = wrapper.find(`#${vidId}`)[0];
//         const canvas = wrapper.find(`#${canvasId}`)[0];
//         const ctx = canvas.getContext("2d");

//         let currentStream = null;
//         let scanning = true;

//         // Stop camera helper
//         function stopCamera() {
//             scanning = false;
//             if (currentStream) {
//                 currentStream.getTracks().forEach(t => t.stop());
//                 currentStream = null;
//             }
//         }

//         // Stop camera when dialog closes
//         const modal = d.$wrapper.find('.modal');
//         modal.on('hidden.bs.modal', () => {
//             stopCamera();
//         });

//         // Request camera access
//         if (!navigator.mediaDevices || !navigator.mediaDevices.getUserMedia) {
//             frappe.msgprint("Camera API not supported in this browser.");
//             return;
//         }

//         navigator.mediaDevices.getUserMedia({ video: { facingMode: "environment" } })
//             .then(stream => {
//                 currentStream = stream;
//                 video.srcObject = stream;
//                 video.play();

//                 // Start scanning loop
//                 const tick = () => {
//                     if (!scanning) return;
//                     if (video.readyState === video.HAVE_ENOUGH_DATA) {
//                         canvas.height = video.videoHeight;
//                         canvas.width = video.videoWidth;
//                         ctx.drawImage(video, 0, 0, canvas.width, canvas.height);
//                         const imageData = ctx.getImageData(0, 0, canvas.width, canvas.height);
//                         const code = jsQR(imageData.data, imageData.width, imageData.height);
//                         if (code) {
//                             frappe.msgprint("Hello");
//                             stopCamera();
//                             d.hide();
//                             return;
//                         }
//                     }
//                     requestAnimationFrame(tick);
//                 };
//                 requestAnimationFrame(tick);
//             })
//             .catch(err => {
//                 frappe.msgprint("Unable to access camera: " + err.message);
//             });
//     });
// }



// opens camera all fine
// function open_camera_dialog(frm) {
//     // Create dialog with an HTML field to hold the <video>
//     let d = new frappe.ui.Dialog({
//         title: "Camera",
//         fields: [
//             { fieldtype: "HTML", fieldname: "camera_area" }
//         ],
//         primary_action_label: "Close",
//         primary_action: function () {
//             // Will trigger bootstrap modal close which we handle below
//             d.hide();
//         }
//     });

//     d.show();

//     // Prepare video element
//     const wrapper = d.get_field("camera_area").$wrapper;
//     const vidId = "se-camera-" + frappe.utils.get_random(6);
//     wrapper.html(`<div style="text-align:center;">
//                     <video id="${vidId}" autoplay playsinline style="max-width:100%; height:auto; border:1px solid #ccc; border-radius:6px;"></video>
//                   </div>
//                   <div style="text-align:center; margin-top:8px;">
//                     <small>Allow camera permission if prompted. Close dialog to stop camera.</small>
//                   </div>`);

//     const video = wrapper.find(`#${vidId}`)[0];
//     let currentStream = null;

//     // Helper to stop camera
//     function stopCamera() {
//         if (currentStream) {
//             currentStream.getTracks().forEach(t => t.stop());
//             currentStream = null;
//         }
//     }

//     // Stop camera when dialog closes (handles both primary action and modal close)
//     const modal = d.$wrapper.find('.modal');
//     modal.on('hidden.bs.modal', () => {
//         stopCamera();
//     });

//     // Request camera
//     if (!navigator.mediaDevices || !navigator.mediaDevices.getUserMedia) {
//         frappe.msgprint("Camera API not supported in this browser.");
//         return;
//     }

//     navigator.mediaDevices.getUserMedia({ video: { facingMode: "environment" } })
//         .then(stream => {
//             currentStream = stream;
//             try {
//                 video.srcObject = stream;
//             } catch (e) {
//                 // fallback
//                 video.src = window.URL.createObjectURL(stream);
//             }
//             video.play().catch(() => { /* play may be blocked until user interacts */ });
//         })
//         .catch(err => {
//             frappe.msgprint("Unable to access camera: " + err.message);
//             // close dialog if you want:
//             // d.hide();
//         });
// }
